### PR TITLE
Add clustering overload for generic metric

### LINKED ISSUE #126: Add clustering overload for user-specified distance metric
### Summary

The `cluster_detections` function currently hardcodes the squared Euclidean distance, but users might want to bring their own metric. We should give them the option to do so.

Related Jira story: [CDAR-700](https://usdot-carma.atlassian.net/browse/CDAR-700)

### Reasoning for new functionality

_No response_

[CDAR-700]: https://usdot-carma.atlassian.net/browse/CDAR-700?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### PR DESCRIPTION
**Pull request by adamlm**
# PR Details
## Description

The `cluster_detections` function used to hardcode the metric type. Now there is an overload for users to bring their own.

## Related GitHub Issue

Closes #126

## Related Jira Key

Closes [CDAR-700](https://usdot-carma.atlassian.net/browse/CDAR-700)

## Motivation and Context

Not all users will want to use scaled Euclidean distance when generating clusters. Previously, the `cluster_detections` function did not allow users to bring their own metrics.

## How Has This Been Tested?

Manually in downstream packages.

## Types of changes

- [x] New feature (non-breaking change that adds functionality)

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.

[CDAR-700]: https://usdot-carma.atlassian.net/browse/CDAR-700?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: include/multiple_object_tracking/clustering.hpp
Comment (paulbourelly999) [inline include/multiple_object_tracking/clustering.hpp:214]: If you are building in some kind of customization point here it might be worth while documenting how someone would bring their own `Metrc` here

Comment (adamlm) [reply include/multiple_object_tracking/clustering.hpp:214]: The whole library needs some documentation overhaul. I can do that now piece-by-piece, or wait until the rest of the current PRs get merged. Do you have a preference?

Comment (paulbourelly999) [reply include/multiple_object_tracking/clustering.hpp:214]: I would say that if possible piece by piece. Ideally each PR would only add documentation related to the pieces of code it touches. We can add some documentation related to how metric can be used here and save other improvements for later. Otherwise I feel that it will be deprioritized.

If you feel that it may be harder to do the documentation piece by piece then maybe we can consider some way to better divide it.

**Review by paulbourelly999 — CHANGES_REQUESTED** [review]
If we can add some documentation here piece-by-piece that would be ideal.

**Review by MishkaMN — APPROVED** [review]
I am good with it if Paul is good with it.

File: include/multiple_object_tracking/clustering.hpp
Comment (paulbourelly999) [reply include/multiple_object_tracking/clustering.hpp:227]: @adamlm  this is good. I was more thinking along the lines of how someone would replace our current metric with a difference custom metric since that is really the change made available by this PR. I am not sure that is currently clear and providing this customization is a lot more valuable with information about how to use it.

**Review by paulbourelly999 — APPROVED** [review]

**Pull request merged**
